apps/prairielearn/src/pages/submissionFile/submissionFile.js

```diff
@@ -27,39 +29,55 @@ async function guessMimeType(name, buffer) {
   return isBinary ? 'application/octet-stream' : 'text/plain';
 }
 
-router.get(
-  '/*',
-  asyncHandler(async (req, res) => {
-    const submissionId = req.params.submission_id;
-    const fileName = req.params[0];
+module.exports = function (options = { publicEndpoint: false }) {
+  const router = express.Router({ mergeParams: true });
 
-    const fileRes = await sqldb.queryZeroOrOneRowAsync(sql.select_submission_file, {
-      question_id: res.locals.question.id,
-      instance_question_id: res.locals.instance_question?.id ?? null,
-      submission_id: submissionId,
-      file_name: fileName,
-    });
+  router.get(
+    '/*',
+    asyncHandler(async (req, res) => {
+      console.log('submission file endpoint hit');
+      if (options.publicEndpoint) {
+        res.locals.course = await selectCourseById(req.params.course_id);
+        res.locals.question = await selectQuestionById(req.params.question_id);
 
-    if (fileRes.rowCount === 0) {
-      res.sendStatus(404);
-      return;
-    }
+        if (
+          !res.locals.question.shared_publicly ||
+          res.locals.course.id !== res.locals.question.course_id
+        ) {
+          throw error.make(404, 'Not Found');
```

Comment:
This should use `res.sendStatus(404)` instead of throwing an error; see our handling of zero rows below. We don't want to render the entire error page in response to these requests.

Reply:
oh right, thanks. All fixed.